### ISSUE TITLE
Missing codec class: groovy.json.internal.LazyMap

### ISSUE THREAD

**Issue by rmorrise** [grails/grails-data-mapping#1082]
### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Summary

My use case (and the corresponding error) are almost identical to:
https://github.com/grails/grails-data-mapping/issues/691

However, the missing codec class is: `groovy.json.internal.LazyMap`. This is the type bound by RestfulController for a POST or PUT request.

### Steps to Reproduce

1. Set up grails project with mongodb plugin (profile=rest-api)
2. Create domain class having an embedded Map property (Sample)
3. Create RestfulController with default implementation (SampleController)
4. POST some test data (http://localhost:8080/Sample) like the following:
```
{
	"name": "Test",
	"value": {
		"foo": {
			"bar": true
		}
	}
}
```

### Expected Behaviour

The domain object should be saved and displayed properly in the response.

### Actual Behaviour

Internal server error.

```
Caused by: org.bson.codecs.configuration.CodecConfigurationException: Can't find a codec for class groovy.json.internal.LazyMap.
	at org.bson.codecs.configuration.CodecCache.getOrThrow(CodecCache.java:46)
	at org.bson.codecs.configuration.ProvidersCodecRegistry.get(ProvidersCodecRegistry.java:63)
	at org.bson.codecs.configuration.ChildCodecRegistry.get(ChildCodecRegistry.java:51)
	at org.grails.datastore.bson.codecs.CodecExtensions$MapCodec.encode(CodecExtensions.groovy:309)
	at org.grails.datastore.bson.codecs.encoders.BasicCollectionTypeEncoder.encode(BasicCollectionTypeEncoder.groovy:43)
	at org.grails.datastore.bson.codecs.BsonPersistentEntityCodec.encode(BsonPersistentEntityCodec.groovy:180)
	at org.grails.datastore.bson.codecs.BsonPersistentEntityCodec.encode(BsonPersistentEntityCodec.groovy:153)
	at org.bson.codecs.BsonDocumentWrapperCodec.encode(BsonDocumentWrapperCodec.java:63)
	at org.bson.codecs.BsonDocumentWrapperCodec.encode(BsonDocumentWrapperCodec.java:29)
	at com.mongodb.connection.InsertCommandMessage.writeTheWrites(InsertCommandMessage.java:101)
	at com.mongodb.connection.InsertCommandMessage.writeTheWrites(InsertCommandMessage.java:43)
	at com.mongodb.connection.BaseWriteCommandMessage.encodeMessageBodyWithMetadata(BaseWriteCommandMessage.java:129)
	at com.mongodb.connection.RequestMessage.encodeWithMetadata(RequestMessage.java:160)
	at com.mongodb.connection.WriteCommandProtocol.sendMessage(WriteCommandProtocol.java:220)
	at com.mongodb.connection.WriteCommandProtocol.execute(WriteCommandProtocol.java:101)
	at com.mongodb.connection.InsertCommandProtocol.execute(InsertCommandProtocol.java:67)
	at com.mongodb.connection.InsertCommandProtocol.execute(InsertCommandProtocol.java:37)
	at com.mongodb.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:168)
	at com.mongodb.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:289)
	at com.mongodb.connection.DefaultServerConnection.insertCommand(DefaultServerConnection.java:118)
	at com.mongodb.operation.MixedBulkWriteOperation$Run$2.executeWriteCommandProtocol(MixedBulkWriteOperation.java:465)
	at com.mongodb.operation.MixedBulkWriteOperation$Run$RunExecutor.execute(MixedBulkWriteOperation.java:656)
	at com.mongodb.operation.MixedBulkWriteOperation$Run.execute(MixedBulkWriteOperation.java:411)
	at com.mongodb.operation.MixedBulkWriteOperation$1.call(MixedBulkWriteOperation.java:177)
	at com.mongodb.operation.MixedBulkWriteOperation$1.call(MixedBulkWriteOperation.java:168)
	at com.mongodb.operation.OperationHelper.withConnectionSource(OperationHelper.java:426)
	at com.mongodb.operation.OperationHelper.withConnection(OperationHelper.java:417)
	at com.mongodb.operation.MixedBulkWriteOperation.execute(MixedBulkWriteOperation.java:168)
	at com.mongodb.operation.MixedBulkWriteOperation.execute(MixedBulkWriteOperation.java:74)
	at com.mongodb.Mongo.execute(Mongo.java:845)
	at com.mongodb.Mongo$2.execute(Mongo.java:828)
	at com.mongodb.MongoCollectionImpl.bulkWrite(MongoCollectionImpl.java:301)
	at com.mongodb.MongoCollectionImpl.bulkWrite(MongoCollectionImpl.java:248)
	at org.grails.datastore.mapping.mongo.MongoCodecSession.flush(MongoCodecSession.groovy:234)
	at org.grails.datastore.mapping.mongo.AbstractMongoSession.flush(AbstractMongoSession.java:74)
	at org.grails.datastore.gorm.GormInstanceApi.doSave(GormInstanceApi.groovy:367)
	at org.grails.datastore.gorm.GormInstanceApi$_save_closure5.doCall(GormInstanceApi.groovy:179)
```

### Environment Information

- **Operating System**: Win7 x64
- **GORM Version:** 6.1.9.RELEASE
- **Grails Version (if using Grails):** 3.3.4
- **JDK Version:** 1.8.0_66
- **Plugin version:** ``` compile 'org.grails.plugins:mongodb:6.1.5' ```

### Example Application

- https://github.com/rmorrise/gorm-mongodb-map-binding

**Comment by rmorrise** [comment]
Will update with steps to reproduce/sample project.

**Comment by rmorrise** [comment]
Updated with complete details and example.

**Comment by rmorrise** [comment]
It looks like this is already fixed in MongoClient in mongodb driver 3.5.0.

**Comment by rmorrise** [comment]
Confirmed workaround for this error - add the new driver to build.gradle:
```compile "org.mongodb:mongodb-driver:3.5.0"```

The map property still doesn't show up in the GSON view, though.
